WordPress/gutenberg
# Style engine: add support for nested CSS rules to `wp_style_engine_get_stylesheet_from_css_rule()`

## What?

Builds on:

- https://github.com/WordPress/gutenberg/pull/58867

by:

1. renaming `$at_rule` to `$rules_group`. 
2. Creates a new class to store rules groups.
3. Allows nested rules to be combined and grouped under similar rules groups when compiled to CSS. 
4. Ensuring that nesting rules groups appear at the end of any compiled CSS. **Note**: this PR does not yet discern between rules groups, e.g., nesting CSS style groups and at-rules
5. Adds missing tests

A related exploratory PR for supporting nested rules is here:

- https://github.com/WordPress/gutenberg/pull/58797

## Why?

1. To make it clear that it supports other styles nesting patterns. For example, "rules group" could be a [nesting style rule](https://www.w3.org/TR/css-nesting-1/#nested-style-rule), a [conditional group rule](https://www.w3.org/TR/css-conditional-3/#conditional-group-rule) or [at-rule groups](https://www.w3.org/TR/css-syntax-3/#at-rule).
2. To separate rules and rules group functionality, e.g., generating each's CSS.
3. Combining declarations is already supported for existing rules. The PR extends that pattern to existing rules groups.
4. There are cases where `@at-rules` don't affect specificity.

## How?
Main changes are the the `WP_Style_Engine_Processor`, which merges incoming/stored rules and outputs compiled CSS with rules groups appearing after any regular CSS rules.

## Potential follow ups in the future

- [ ] Add the same functionality to `wp_style_engine_get_styles()`
- [ ] At the moment `"$rules_group . $selector"` keys are stored in the store, which means they can be overwritten. Later we can allow to add and merge nested rules to the store.
- [ ] Discern between rules groups, e.g., nesting CSS style groups and at-rules when considering order
- [ ] Allow rules groups to have top-level CSS declarations, e.g., `.ham { color: red; &:hover { color: blue; } }`
- [ ] JS version

## Testing Instructions

Run the tests

`npm run test:unit:php:base -- --group style-engine`

Fire up this branch using a block theme, e.g., 2024. Check that the `core-block-supports-inline-css` on the frontend is exactly the same as trunk.

You can manually check how rules are generated and printed. 

<details>

<summary>Here are some rules declared by gutenberg_style_engine_get_stylesheet_from_css_rules()</summary>

```php
	gutenberg_style_engine_get_stylesheet_from_css_rules(
		array(
			array(
				'rules_group'  => '@media (orientation: landscape)',
				'selector'     => '.ham',
				'declarations' => array( 'gap' => '10px' ),
			)
		),
		array(
			'context'  => 'block-supports',
			'prettify' => true,
		)
	);

	gutenberg_style_engine_get_stylesheet_from_css_rules(
		array(
			array(
				'rules_group'  => '@media (orientation: landscape)',
				'selector'     => '.ham',
				'declarations' => array( 'gap' => '1px', 'background-color' => 'blue' ),
			)
		),
		array(
			'context'  => 'block-supports',
			'prettify' => true,
		)
	);

	gutenberg_style_engine_get_stylesheet_from_css_rules(
		array(
			array(
				'rules_group'  => '@media (orientation: landscape)',
				'selector'     => '.jam',
				'declarations' => array( 'background-color' => 'pink' ),
			)
		),
		array(
			'context'  => 'block-supports',
			'prettify' => true,
		)
	);

	gutenberg_style_engine_get_stylesheet_from_css_rules(
		array(
			array(
				'rules_group'  => '@layer',
				'selector'     => '.toast',
				'declarations' => array( 'color' => 'red' ),
			)
		),
		array(
			'context'  => 'block-supports',
			'prettify' => true,
		)
	);

	gutenberg_style_engine_get_stylesheet_from_css_rules(
		array(
			array(
				'rules_group'  => '@layer',
				'selector'     => '.toast:hover',
				'declarations' => array( 'color' => 'purple' ),
			)
		),
		array(
			'context'  => 'block-supports',
			'prettify' => true,
		)
	);


```

</details>


The CSS should be printed at the end of `core-block-supports-inline-css` rules stylesheet:

``` css
@media (orientation: landscape) {
	.ham {
		gap: 1px;
		background-color: blue;
	}
	.jam {
		background-color: pink;
	}
}
@layer {
	toast {
		color: red;
	}
	toast:hover {
		color: purple;
	}
}
```
